source/common/router/router.cc

```diff
@@ -2030,15 +2098,50 @@ void Filter::doRetry(bool can_send_early_data, bool can_use_http3, TimeoutRetry
   callbacks_->streamInfo().setAttemptCount(attempt_count_);
   ASSERT(pending_retries_ > 0);
   pending_retries_--;
+  if (host_selection_cancelable_) {
+    // If there was a timeout during host selection, cancel this selection
+    // attempt before potentially creating a new one.
+    host_selection_cancelable_->cancel();
+    host_selection_cancelable_.reset();
+  }
 
   // Clusters can technically get removed by CDS during a retry. Make sure it still exists.
   const auto cluster = config_->cm_.getThreadLocalCluster(route_entry_->clusterName());
   std::unique_ptr<GenericConnPool> generic_conn_pool;
-  if (cluster != nullptr) {
-    cluster_ = cluster->info();
-    generic_conn_pool = createConnPool(*cluster);
+  if (cluster == nullptr) {
```

Comment:
In the old code it looks like this is guarded by `if (!generic_conn_pool)` Do we need this here?

Reply:
I split the noHealthyUpstream into two places - one where we have no cluster, and one where we fail to get a host.  Trying to keep it in one place was problematic given that host selection can happen later